File: apps/ios/MelodyLien/ViewModels/AppViewModel.swift
import SwiftUI
import Combine

// MARK: - Navigation

enum Screen {
    case home
    case exchange
    case mystery(trackId: String)
    case puzzle(trackId: String)
    case artist(artistId: String)
    case playlist
}

// MARK: - AppViewModel

@MainActor
final class AppViewModel: ObservableObject {

    // MARK: Published

    @Published var user: User?
    @Published var tracks: [String: Track] = [:]       // trackId → Track
    @Published var collection: CollectionSummary?
    @Published var mission: Mission?
    @Published var todayEncounter: Encounter?
    @Published var dailyPlaylist: DailyPlaylist?

    @Published var activeTab: Tab = .home
    @Published var navigationStack: [Screen] = []

    @Published var selectedCandidateIndex: Int? = nil
    @Published var isLoading = false
    @Published var errorMessage: String? = nil
    @Published var toast: String? = nil

    // MARK: Dependencies

    let bleManager: BLEManager
    private let repository: MelodyRepository

    // MARK: Init

    init(repository: MelodyRepository = .shared) {
        self.repository = repository
        self.bleManager = BLEManager(repository: repository)
    }

    // MARK: - Load

    func loadAll() async {
        isLoading = true
        defer { isLoading = false }
        await withTaskGroup(of: Void.self) { group in
            group.addTask { await self.loadUser() }
            group.addTask { await self.loadTracks() }
            group.addTask { await self.loadCollection() }
            group.addTask { await self.loadMission() }
            group.addTask { await self.loadEncounter() }
            group.addTask { await self.loadPlaylist() }
        }
    }

    private func loadUser() async {
        do { user = try await repository.fetchMe() }
        catch { handle(error) }
    }

    private func loadTracks() async {
        do {
            let list = try await repository.fetchTracks()
            tracks = Dictionary(uniqueKeysWithValues: list.map { ($0.id, $0) })
        } catch { handle(error) }
    }

    private func loadCollection() async {
        do { collection = try await repository.fetchCollection() }
        catch { handle(error) }
    }

    private func loadMission() async {
        do { mission = try await repository.fetchMission() }
        catch { handle(error) }
    }

    private func loadEncounter() async {
        do { todayEncounter = try await repository.fetchTodayEncounter() }
        catch { handle(error) }
    }

    private func loadPlaylist() async {
        do { dailyPlaylist = try await repository.fetchDailyPlaylist() }
        catch { handle(error) }
    }

    // MARK: - Actions

    func selectCandidate(_ index: Int) {
        selectedCandidateIndex = index
    }

    func confirmPiece() async {
        guard let encounter = todayEncounter,
              let index = selectedCandidateIndex else { return }

        do {
            let result = try await repository.selectPiece(encounter: encounter, candidateIndex: index)
            // ローカルのtrack状態を更新
            if var track = tracks[result.trackId] {
                if result.added && !track.ownedPieces.contains(result.pieceNumber) {
                    track.ownedPieces.append(result.pieceNumber)
                    track.ownedPieces.sort()
                    tracks[result.trackId] = track
                }
            }
            selectedCandidateIndex = nil
            showToast(result.added ? "ピースを獲得しました！" : "所持済みのピースです")
            // 画面遷移
            if result.nextScreen == "puzzle" {
                navigationStack.append(.puzzle(trackId: result.trackId))
            } else {
                navigationStack.append(.mystery(trackId: result.trackId))
            }
        } catch { handle(error) }
    }

    func applyHint(trackId: String, kind: HintKind) async {
        do {
            let result = try await repository.applyHint(trackId: trackId, kind: kind)
            if var track = tracks[trackId] {
                track.hintLevel    = result.hintLevel
                track.answerReady  = result.answerReady
                tracks[trackId]    = track
            }
            showToast(kind == .answer ? "YouTubeリンクを表示しました" : "ヒントを解放しました")
        } catch { handle(error) }
    }

    func unlockTrack(trackId: String) async {
        do {
            try await repository.unlockTrack(trackId: trackId)
            if var track = tracks[trackId] {
                track.isUnlocked = true
                tracks[trackId]  = track
            }
            showToast("曲が解放されました！")
            navigationStack.removeLast()
            navigationStack.append(.puzzle(trackId: trackId))
        } catch { handle(error) }
    }

    func addListenLater(trackId: String) async {
        do {
            try await repository.addListenLater(trackId: trackId)
            showToast("あとで聴くに追加しました")
        } catch { handle(error) }
    }

    func navigate(to screen: Screen) {
        navigationStack.append(screen)
    }

    func goBack() {
        navigationStack.removeLast()
    }

    // MARK: - Computed

    var heroTrack: Track? {
        guard let playlist = dailyPlaylist,
              let item = playlist.tracks.first else { return nil }
        return tracks[item.trackId]
    }

    func track(for id: String) -> Track? { tracks[id] }

    // MARK: - Helpers

    private func showToast(_ message: String) {
        toast = message
        Task {
            try? await Task.sleep(nanoseconds: 2_500_000_000)
            toast = nil
        }
    }

    private func handle(_ error: Error) {
        errorMessage = error.localizedDescription
    }
}

// MARK: - Tab

enum Tab: String, CaseIterable {
    case home      = "ホーム"
    case exchange  = "届いた"
    case puzzle    = "パズル"
    case artist    = "アーティスト"
    case playlist  = "プレイリスト"

    var icon: String {
        switch self {
        case .home:     return "house.fill"
        case .exchange: return "arrow.left.arrow.right.circle.fill"
        case .puzzle:   return "puzzlepiece.fill"
        case .artist:   return "star.fill"
        case .playlist: return "music.note.list"
        }
    }
}

File: apps/android/app/src/main/kotlin/com/streetmelody/app/viewmodel/AppViewModel.kt
package com.streetmelody.app.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.streetmelody.app.data.model.*
import com.streetmelody.app.data.repository.MelodyRepository
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

// MARK: - Navigation

sealed class Screen {
    data object Home      : Screen()
    data object Exchange  : Screen()
    data class Mystery(val trackId: String) : Screen()
    data class Puzzle(val trackId: String)  : Screen()
    data class Artist(val artistId: String) : Screen()
    data object Playlist  : Screen()
}

enum class Tab { HOME, EXCHANGE, PUZZLE, ARTIST, PLAYLIST }

// MARK: - UI State

data class HomeUiState(
    val user: User? = null,
    val collection: CollectionSummary? = null,
    val mission: Mission? = null,
    val heroTrack: Track? = null,
    val recentTracks: List<Track> = emptyList(),
)

// MARK: - AppViewModel

class AppViewModel(app: Application) : AndroidViewModel(app) {

    private val repository = MelodyRepository()

    // Nav
    private val _activeTab = MutableStateFlow(Tab.HOME)
    val activeTab: StateFlow<Tab> = _activeTab

    private val _navStack = MutableStateFlow<List<Screen>>(emptyList())
    val navStack: StateFlow<List<Screen>> = _navStack

    // Data
    private val _tracks = MutableStateFlow<Map<String, Track>>(emptyMap())
    val tracks: StateFlow<Map<String, Track>> = _tracks

    private val _encounter = MutableStateFlow<Encounter?>(null)
    val encounter: StateFlow<Encounter?> = _encounter

    private val _playlist = MutableStateFlow<DailyPlaylist?>(null)
    val playlist: StateFlow<DailyPlaylist?> = _playlist

    val homeState: StateFlow<HomeUiState> = combine(
        _tracks, _encounter, _playlist
    ) { tracks, _, playlist ->
        val heroTrackId = playlist?.tracks?.firstOrNull()?.trackId
        HomeUiState(
            heroTrack = heroTrackId?.let { tracks[it] },
            recentTracks = tracks.values.take(3).toList(),
        )
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), HomeUiState())

    // UI state
    private val _selectedCandidate = MutableStateFlow<Int?>(null)
    val selectedCandidate: StateFlow<Int?> = _selectedCandidate

    private val _toast = MutableStateFlow<String?>(null)
    val toast: StateFlow<String?> = _toast

    private val _user = MutableStateFlow<User?>(null)
    val user: StateFlow<User?> = _user

    private val _collection = MutableStateFlow<CollectionSummary?>(null)
    val collection: StateFlow<CollectionSummary?> = _collection

    private val _mission = MutableStateFlow<Mission?>(null)
    val mission: StateFlow<Mission?> = _mission

    init { loadAll() }

    // MARK: - Load

    fun loadAll() {
        viewModelScope.launch {
            launch { runCatching { _user.value = repository.fetchMe() } }
            launch { runCatching { _collection.value = repository.fetchCollection() } }
            launch { runCatching { _mission.value = repository.fetchMission() } }
            launch { runCatching {
                val list = repository.fetchTracks()
                _tracks.value = list.associateBy { it.id }
            }}
            launch { runCatching { _encounter.value = repository.fetchTodayEncounter() } }
            launch { runCatching { _playlist.value = repository.fetchDailyPlaylist() } }
        }
    }

    // MARK: - Actions

    fun selectCandidate(index: Int) { _selectedCandidate.value = index }

    fun confirmPiece() {
        val enc = _encounter.value ?: return
        val idx = _selectedCandidate.value ?: return
        viewModelScope.launch {
            runCatching { repository.selectPiece(enc.id, idx) }.onSuccess { result ->
                _tracks.value = _tracks.value.toMutableMap().also { map ->
                    map[result.trackId]?.let { track ->
                        if (result.added && !track.ownedPieces.contains(result.pieceNumber)) {
                            map[result.trackId] = track.copy(ownedPieces = (track.ownedPieces + result.pieceNumber).sorted())
                        }
                    }
                }
                _selectedCandidate.value = null
                showToast(if (result.added) "ピースを獲得しました！" else "所持済みのピースです")
                val screen = if (result.nextScreen == "puzzle") Screen.Puzzle(result.trackId)
                             else Screen.Mystery(result.trackId)
                _navStack.value = _navStack.value + screen
            }
        }
    }

    fun applyHint(trackId: String, kind: HintKind) {
        viewModelScope.launch {
            runCatching { repository.applyHint(trackId, kind) }.onSuccess { result ->
                _tracks.value = _tracks.value.toMutableMap().also { map ->
                    map[trackId]?.let {
                        map[trackId] = it.copy(hintLevel = result.hintLevel, answerReady = result.answerReady)
                    }
                }
                showToast(if (kind == HintKind.ANSWER) "YouTubeリンクを表示しました" else "ヒントを解放しました")
            }
        }
    }

    fun unlockTrack(trackId: String) {
        viewModelScope.launch {
            runCatching { repository.unlockTrack(trackId) }.onSuccess {
                _tracks.value = _tracks.value.toMutableMap().also { map ->
                    map[trackId]?.let { map[trackId] = it.copy(isUnlocked = true) }
                }
                showToast("曲が解放されました！")
                val newStack = _navStack.value.dropLast(1) + Screen.Puzzle(trackId)
                _navStack.value = newStack
            }
        }
    }

    fun navigate(screen: Screen) { _navStack.value = _navStack.value + screen }
    fun goBack()                  { _navStack.value = _navStack.value.dropLast(1) }
    fun setTab(tab: Tab)          { _activeTab.value = tab }

    private fun showToast(msg: String) {
        _toast.value = msg
        viewModelScope.launch {
            kotlinx.coroutines.delay(2500)
            _toast.value = null
        }
    }
}

enum class HintKind { HINT1, HINT2, ANSWER }
